backends/gstreamer/datachannel.rs

```diff
@@ -45,83 +57,96 @@ impl GStreamerWebRtcDataChannel {
             .map_err(|e| e.to_string())?
             .expect("Invalid datachannel");
 
-        GStreamerWebRtcDataChannel::from(channel)
+        GStreamerWebRtcDataChannel::from(id, channel, thread)
     }
 
-    pub fn from(channel: glib::Object) -> Result<Self, String> {
-        let callbacks = Arc::new(Mutex::new(WebRtcDataChannelCallbacks::new()));
-
-        let callbacks_ = callbacks.clone();
+    pub fn from(
+        id: &DataChannelId,
+        channel: glib::Object,
+        thread: &WebRtcThread,
+    ) -> Result<Self, String> {
+        let id_ = id.clone();
+        let thread_ = Mutex::new(thread.clone());
         channel
             .connect("on-open", false, move |_| {
-                callbacks_.lock().unwrap().open();
+                thread_
+                    .lock()
+                    .unwrap()
+                    .internal_event(InternalEvent::OnDataChannelEvent(
+                        id_,
+                        DataChannelEvent::Open,
+                    ));
                 None
             })
             .map_err(|e| e.to_string())?;
 
-        let callbacks_ = callbacks.clone();
+        let id_ = id.clone();
+        let thread_ = Mutex::new(thread.clone());
         channel
             .connect("on-close", false, move |_| {
-                callbacks_.lock().unwrap().close();
+                thread_
+                    .lock()
+                    .unwrap()
+                    .internal_event(InternalEvent::OnDataChannelEvent(
+                        id_,
+                        DataChannelEvent::Close,
+                    ));
                 None
             })
             .map_err(|e| e.to_string())?;
 
-        let callbacks_ = callbacks.clone();
+        let id_ = id.clone();
+        let thread_ = Mutex::new(thread.clone());
         channel
             .connect("on-error", false, move |error| {
                 if let Some(error) = error[0]
                     .get::<glib::error::Error>()
                     .expect("Invalid GError")
                 {
-                    callbacks_
+                    thread_
                         .lock()
                         .unwrap()
-                        .error(WebRtcError::Backend(error.to_string()));
+                        .internal_event(InternalEvent::OnDataChannelEvent(
+                            id_,
+                            DataChannelEvent::Error(WebRtcError::Backend(error.to_string())),
+                        ));
                 }
                 None
             })
             .map_err(|e| e.to_string())?;
 
-        let callbacks_ = callbacks.clone();
+        let id_ = id.clone();
+        let thread_ = Mutex::new(thread.clone());
```

Comment:
the outer mutex doesn't seem necessary?

Reply:
@sdroege should the `on-message-string` and `on-error` callbacks require Sync here? I recall we fixed some of these cases earlier: https://github.com/sdroege/gstreamer-rs/issues/154

Reply:
`rustc` complains if I don't add the outer mutex

Reply:
Yeah sorry this is the Sync requirement that I think needs to be removed upstream. No worries for now